### PR TITLE
Add more coercions for table/partition mismatch issue in Hive connector

### PR DESCRIPTION
**Pull request by hustnn**
Trying to address https://github.com/prestosql/presto/issues/137

To do:
- [x] decimal to decimal.
- [x] decimal and double convert.
- [x] decimal and float convert.
- [x] double to float.

<del> Currently only support HivePageSource, add support for RecordPageSource.

Updated: this PR supports HivePageSource first.

### CODE REVIEW

**Comment by dain** [comment]
@hustnn in the description above, I see `add support for RecordPageSource`, does this mean that this a work in progress?  If so, it it is helpful to the reviewers if you can "edit" the PR title to start with `[WIP]`, so they know it isn't necessary to review yet.

**Comment by hustnn** [comment]
@dain Hi, I found GenericHiveRecordCursorProvider automatically do the coercion without HiveCoercionRecordCursor.

**Comment by dain** [comment]
@hustnn, interesting, looks like `S3SelectRecordCursorProvider` is the only code using the adapter now.  So is this ready to review?

**Comment by hustnn** [comment]
@dain . I will add float and decimal, double and decimal coercion. Then, should be ok for review. For this PR, can we finish the support for hive page source first ?

For the RecordCursor, If also needed, I can open another PR.

**Comment by hustnn** [comment]
@dain Ready for review.

**Comment by hustnn** [comment]
> Some super minor comments. This will need review from some that really understands decimal.

Minor comments addressed.

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java [hunk 1]
```diff
@@ -106,6 +110,11 @@
                         "    int_to_bigint              INT," +
                         "    bigint_to_varchar          BIGINT," +
                         "    float_to_double            " + floatToDoubleType + "," +
+                        "    double_to_float            DOUBLE," +
+                        "    shortdecimal_to_shortdecimal          DECIMAL(10,2)," +
+                        "    shortdecimal_to_longdecimal          DECIMAL(10,2)," +
+                        "    longdecimal_to_shortdecimal          DECIMAL(20,12)," +
+                        "    longdecimal_to_longdecimal          DECIMAL(20,12)," +
```
Comment (sopel39) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java]: What about `float to decimal`, `double to decimal`, `decimal to float`, `decimal to double`?

**Comment by dain** [comment]
> * a lot of code has been copied from `presto-main`. I wonder if it qualifies for a separate module. What do you think @dain @electrum ?

That seems reasonable, but I don't think we should delay this PR for that reorganization.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+        public ShortDecimalToShortDecimalCoercer(DecimalType fromType, DecimalType toType)
+        {
+            super(fromType, toType);
+            rescale = longTenToNth(Math.abs(toType.getScale() - fromType.getScale()));
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            long returnValue = shortToShortCast(fromType.getLong(block, position),
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java]: nit: each argument in new line

File: presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+    {
+        public ShortDecimalToLongDecimalCoercer(DecimalType fromType, DecimalType toType)
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            Slice coercedValue = shortToLongCast(fromType.getLong(block, position),
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java]: nit: each argument in new line

File: presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+    {
+        public LongDecimalToShortDecimalCoercer(DecimalType fromType, DecimalType toType)
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            long returnValue = longToShortCast(fromType.getSlice(block, position),
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java]: ditto

File: presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+    {
+        public LongDecimalToLongDecimalCoercer(DecimalType fromType, DecimalType toType)
+        {
+            super(fromType, toType);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            Slice coercedValue = longToLongCast(fromType.getSlice(block, position),
```
Comment (sopel39) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java]: ditto: each argument in new line

File: presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java [hunk 6]
```diff
@@ -192,26 +232,349 @@ private static Slice longToLongCast(
...
+        public ShortDecimalToDoubleCoercer(DecimalType fromType)
+        {
+            super(fromType, DOUBLE);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            toType.writeDouble(blockBuilder,
+                    shortDecimalToDouble(fromType.getLong(block, position), longTenToNth(fromType.getScale())));
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java]: ping, make `longTenToNth(fromType.getScale()))` field

File: presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java [hunk 7]
```diff
@@ -192,26 +232,349 @@ private static Slice longToLongCast(
...
+        public ShortDecimalToRealCoercer(DecimalType fromType)
+        {
+            super(fromType, REAL);
+        }
+
+        @Override
+        protected void applyCoercedValue(BlockBuilder blockBuilder, Block block, int position)
+        {
+            toType.writeLong(blockBuilder,
+                    shortDecimalToReal(fromType.getLong(block, position), longTenToNth(fromType.getScale())));
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/coercions/DecimalCoercers.java]: ping: make `longTenToNth(fromType.getScale())` a field

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java [hunk 8]
```diff
@@ -13,14 +13,17 @@
  */
 package io.prestosql.plugin.hive;
 
-import io.airlift.slice.Slice;
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java]: ping, commit message is too long

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java (hunk 1 above)
Comment (sopel39) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java]: Will you implement tests for those conversions too?

Comment (hustnn) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java]: @sopel39 Yes, I am working on it. Already labelled WIP. Once finishing, will remove WIP.

Comment (hustnn) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveCoercion.java]: I am fixing all your previous comments.

**Comment by hustnn** [comment]
@sopel39 @dain Comments all fixed.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java [hunk 9]
```diff
@@ -485,6 +499,229 @@ public Block apply(Block block)
...
+                }
+                Slice coercedValue = longToLongCast(fromType.getSlice(block, i), fromType.getPrecision(),
+                        fromType.getScale(), toType.getPrecision(), toType.getScale());
+                toType.writeSlice(blockBuilder, coercedValue);
+            }
+            return blockBuilder.build();
+        }
+    }
+
+    private static long shortToShortCast(
```
[the remark this answers is not in the document]
Comment (dain) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HivePageSource.java]: Yes. I don't want 2 versions of this complex stuff around.

**Comment by dain** [comment]
I'll reorganize the cast code

**Pull request merged**